CASSANDRA-17668 handle leak of non-standard Java types as clients using JMX cannot handle them

lmtrombone opened the pull request:
DatabaseDescriptor had changes where some methods threw non-standard exceptions.  This caused some of the JMX setter methods that called them to throw them as well which causes issues as the clients can't handle them.  This pull request addresses these issues by creating new APIs that catches and converts the non-standard exceptions to standard ones and deprecates the old setters.  For consistency, since the setters have been deprecated, new getters have been created as well.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/ActiveRepairServiceMBean.java:
```diff
@@ -34,9 +34,14 @@
 
+    /**
+     * @deprecated use setRepairSessionSpaceInMiB instead as it will not throw non-standard exceptions
+     */
```
We need to add @Deprecated here too, not only in Java doc and the actual implementation

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4525,6 +4525,7 @@ public List<String> getParentRepairStatus(int cmd)
                ImmutableList.<String>builder().add(pair.left.name()).addAll(pair.right).build();
     }
 
+    @Deprecated
```
I think @Override was missed before and we can use the opportunity to add it maybe?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageServiceMBean.java:
```diff
@@ -835,13 +838,23 @@ default int upgradeSSTables(String keyspaceName, boolean excludeCurrentVersion,
+     * @deprecated use setColumnIndexSizeInKiB instead as it will not throw non-standard exceptions
+     * Sets the granularity of the collation index of rows within a partition
+     */
     public void setColumnIndexSize(int columnIndexSizeInKB);
```
Same as before, let's add also @Deprecated

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageServiceMBean.java:
```diff
@@ -835,13 +838,23 @@ default int upgradeSSTables(String keyspaceName, boolean excludeCurrentVersion,
+    /**
+     * @deprecated use setColumnIndexCacheSizeInKiB instead as it will not throw non-standard exceptions
+     * Sets the threshold for skipping the column index when caching partition info
+     */
```
@Deprecated to be added

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageServiceMBean.java:
```diff
@@ -850,8 +863,13 @@ default int upgradeSSTables(String keyspaceName, boolean excludeCurrentVersion,
+    /**
+     * @deprecated use setBatchSizeWarnThresholdInKiB instead as it will not throw non-standard exceptions
+     * Sets the threshold for warning queries due to a large batch size
+     */
```
Same @Deprecated to be added

ekaterinadimitrova2 reviewed (commented):
Overall looks good on a quick skim, I left some tiny suggestions. Just to confirm, did you have the chance to verify whether there are other setters/getters in the DD that can suffer from this problem, out of the list/examples suggested on the ticket? 
Also, I would like to suggest to add some basic unit tests to DatabaseDescriptorTest class? 
Also, I think we need patches for previous versions and we normally merge them from lower affected version upwards to trunk - for more info https://cassandra.apache.org/_/development/patches.html (Bug Fixes section). There might be also the case we need different patches for the different versions as things diverged between them. I can run a full CI for you when we get things into shape.
One more thing is to check whether our docs need some relevant update, at minimum we need to update the Deprecation section in NEWS.txt. Thank you for your work!

dcapwell commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4525,16 +4525,39 @@ public List<String> getParentRepairStatus(int cmd)
+    }
+
+    @Override
+    public int getRepairSessionMaximumTreeDepth()
```
can you add a comment explaining why this exists?  Fo someone reviewing the code its going to be asked "why is getRepairSessionMaxTreeDepth deprecated... its the same logic here...", so if we are going to depreciate the `get` we should explain why in the code

ekaterinadimitrova2 replied:
Good point, we can also add the ticket number for clarity

dcapwell commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -6146,29 +6169,71 @@ public void setCachedReplicaRowsFailThreshold(int threshold)
         logger.info("updated replica_filtering_protection.cached_rows_fail_threshold to {}", threshold);
     }
 
+    @Override
```
same comment as above, can we document "why" the new method that has identical logic?

lmtrombone replied:
this is actually an existing method that I added `@Override` to for consistency purposes / best practices.  If you still think it makes sense to add a comment here, let me know.

dcapwell commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -6146,29 +6169,71 @@ public void setCachedReplicaRowsFailThreshold(int threshold)
         logger.info("Updated column_index_cache_size to {}", cacheSizeInKB);
     }
 
+    @Override
```
same comment as above, can we document "why" the new method that has identical logic?

dcapwell commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -6180,17 +6245,42 @@ public void setBatchSizeFailureThreshold(int threshold)
     }
 
+    @Override
+    public int getBatchSizeWarnThresholdInKiB()
```
same comment as above, can we document "why" the new method that has identical logic?

lmtrombone committed:
handle leak of non-standard Java types as clients using JMX cannot handle them

-deprecate and replace JMX setters that throw non-standard exceptions
-deprecate and replace respective JMX getters as well to make JMX usage consistent

patch by Leonard Ma; reviewed Ekaterina Dimitrova and David Capwell for CASSANDRA-17668

ekiauhce commented on src/java/org/apache/cassandra/service/ActiveRepairService.java:
```diff
@@ -297,18 +298,44 @@ public int getRepairSessionSpaceInMegabytes()
+        }
+        catch (ConfigurationException e)
+        {
+            throw new IllegalArgumentException(e.getMessage());
```
It's better to use `IllegalArgumentException(Throwable cause)` constructor, because it will preserve `e` stacktrace.

ekaterinadimitrova2 replied:
This would be a precedent here I think, I can't imagine putting the Stacktrace as an error message in JConsole, @dcapwell ?

lmtrombone replied:
Unless I'm missing something, I don't think this will be applicable here.  I tried it and JConsole did not respond well to it and had trouble properly displaying the stacktrace as is.
